Unable to use @Option with ListProperty/SetProperty

### Expected Behavior
`@Option` should be possible to use with `final ListProperty/SetProperty<>` fields (as it was implemented by @adammurdoch for `Property<>` in https://github.com/gradle/gradle/pull/7977).

### Current Behavior
```
@Input
@Optional
@Option(option = "pitest-features", description = "Configures PIT features to be used")
final ListProperty<String> features //it's Groovy, getter is automatically generated
```
gives:
```
OptionValidationException: No setter for Option annotated field 'features' in class 'class info.solidsoft.gradle.pitest.PitestTask'.
```

### Context
After migration from `List<>` and conventionMapping to `PropertyList<>` (or `PropertySet<>`) it is no longer possible to use `@Option` to set/override values.

### Steps to Reproduce 
Try to set value with `@Option` for `final ListProperty<>`/`final SetProperty<>` fileds.

### Your Environment
Gradle 5.6.1.

Looks like the behavior changed somewhat, but it still doesn't work as expected. From duplicate issue https://github.com/gradle/gradle/issues/25097:

### Expected Behavior

It is possible to annotation a JavaBean property of type `List<String>` with `@Option` and supply it with values from the command-line. We should have feature parity for properties of type `ListProperty<String>` etc.

### Current Behavior (optional)

Currently an error is thrown if `@Option` is used on `ListProperty<String>`:

```text
Caused by: org.gradle.cli.CommandLineArgumentException: Command-line option '--serial' does not take an argument.
	at app//org.gradle.cli.CommandLineParser$KnownOptionParserState.onArgument(CommandLineParser.java:423)
	at app//org.gradle.cli.CommandLineParser.parse(CommandLineParser.java:93)
	at app//org.gradle.execution.commandline.CommandLineTaskConfigurer.configureTasksNow(CommandLineTaskConfigurer.java:67)
	... 123 more
```

### Context

When migrating eager multi-valued properties from `List<T>` to lazy `ListProperty<T>` types we need to keep `@Option` working for them.

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/options/AbstractOptionIntegrationSpec.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/options/AbstractOptionIntegrationSpec.groovy
@@ -90,118 +90,24 @@ abstract class AbstractOptionIntegrationSpec extends AbstractIntegrationSpec {
         """
     }
 
-    String taskWithSinglePropertyOption(String optionType) {
-        """
-            import org.gradle.api.DefaultTask;
-            import org.gradle.api.tasks.Internal;
-            import org.gradle.api.tasks.TaskAction;
-            import org.gradle.api.tasks.options.Option;
-            import org.gradle.api.provider.Property;
+    String taskWithSinglePropertyOption(String propertyType, String optionType) {
+        String methodName = propertyType.substring(0, 1).toLowerCase() + propertyType.substring(1)
 
-            import java.util.List;
-
-            public class SampleTask extends DefaultTask {
-                private final Property<$optionType> myProp = getProject().getObjects().property(${optionType}.class);
-
-                @Internal
-                @Option(option = "myProp", description = "Configures command line option 'myProp'.")
-                public Property<$optionType> getMyProp() {
-                    return myProp;
-                }
-
-                @TaskAction
-                public void renderOptionValue() {
-                    System.out.println("Value of myProp: " + myProp.getOrNull());
-                }
-
-                private static enum TestEnum {
-                    OPT_1, OPT_2, OPT_3
-                }
-            }
-        """
-    }
-
-    String groovyTaskWithSinglePropertyOption(String optionType) {
-        """
-            public class SampleTask extends DefaultTask {
-                @Internal
-                @Option(option = "myProp", description = "Configures command line option 'myProp'.")
-                final Property<$optionType> myProp = project.objects.property($optionType)
-
-                @TaskAction
-                public void renderOptionValue() {
-                    println("Value of myProp: " + myProp.getOrNull())
-                }
-
-                private static enum TestEnum {
-                    OPT_1, OPT_2, OPT_3
-                }
-            }
-        """
-    }
-
-    String taskWithListPropertyOption(String optionType) {
         """
             import org.gradle.api.DefaultTask;
             import org.gradle.api.tasks.Internal;
             import org.gradle.api.tasks.TaskAction;
             import org.gradle.api.tasks.options.Option;
             import org.gradle.api.provider.ListProperty;
-
-            public class SampleTask extends DefaultTask {
-                private final ListProperty<$optionType> myProp = getProject().getObjects().listProperty(${optionType}.class);
-
-                @Internal
-                @Option(option = "myProp", description = "Configures command line option 'myProp'.")
-                public ListProperty<$optionType> getMyProp() {
-                    return myProp;
-                }
-
-                @TaskAction
-                public void renderOptionValue() {
-                    System.out.println("Value of myProp: " + myProp.getOrNull());
-                }
-
-                private static enum TestEnum {
-                    OPT_1, OPT_2, OPT_3
-                }
-            }
-        """
-    }
-
-    String groovyTaskWithListPropertyOption(String optionType) {
-        """
-            public class SampleTask extends DefaultTask {
-                @Internal
-                @Option(option = "myProp", description = "Configures command line option 'myProp'.")
-                final ListProperty<$optionType> myProp = project.objects.listProperty($optionType)
-
-                @TaskAction
-                public void renderOptionValue() {
-                    println("Value of myProp: " + myProp.getOrNull())
-                }
-
-                private static enum TestEnum {
-                    OPT_1, OPT_2, OPT_3
-                }
-            }
-        """
-    }
-
-    String taskWithSetPropertyOption(String optionType) {
-        """
-            import org.gradle.api.DefaultTask;
-            import org.gradle.api.tasks.Internal;
-            import org.gradle.api.tasks.TaskAction;
-            import org.gradle.api.tasks.options.Option;
+            import org.gradle.api.provider.Property;
             import org.gradle.api.provider.SetProperty;
 
             public class SampleTask extends DefaultTask {
-                private final SetProperty<$optionType> myProp = getProject().getObjects().setProperty(${optionType}.class);
+                private final $propertyType<$optionType> myProp = getProject().getObjects().${methodName}(${optionType}.class);
 
                 @Internal
                 @Option(option = "myProp", description = "Configures command line option 'myProp'.")
-                public SetProperty<$optionType> getMyProp() {
+                public $propertyType<$optionType> getMyProp() {
                     return myProp;
                 }
 
@@ -217,12 +123,14 @@ abstract class AbstractOptionIntegrationSpec extends AbstractIntegrationSpec {
         """
     }
 
-    String groovyTaskWithSetPropertyOption(String optionType) {
+    String groovyTaskWithSinglePropertyOption(String propertyType, String optionType) {
+        String methodName = propertyType.substring(0, 1).toLowerCase() + propertyType.substring(1)
+
         """
             public class SampleTask extends DefaultTask {
                 @Internal
                 @Option(option = "myProp", description = "Configures command line option 'myProp'.")
-                final SetProperty<$optionType> myProp = project.objects.setProperty($optionType)
+                final $propertyType<$optionType> myProp = project.objects.$methodName($optionType)
 
                 @TaskAction
                 public void renderOptionValue() {

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/options/AbstractOptionIntegrationSpec.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/options/AbstractOptionIntegrationSpec.groovy
@@ -140,6 +140,102 @@ abstract class AbstractOptionIntegrationSpec extends AbstractIntegrationSpec {
         """
     }
 
+    String taskWithListPropertyOption(String optionType) {
+        """
+            import org.gradle.api.DefaultTask;
+            import org.gradle.api.tasks.Internal;
+            import org.gradle.api.tasks.TaskAction;
+            import org.gradle.api.tasks.options.Option;
+            import org.gradle.api.provider.ListProperty;
+
+            public class SampleTask extends DefaultTask {
+                private final ListProperty<$optionType> myProp = getProject().getObjects().listProperty(${optionType}.class);
+
+                @Internal
+                @Option(option = "myProp", description = "Configures command line option 'myProp'.")
+                public ListProperty<$optionType> getMyProp() {
+                    return myProp;
+                }
+
+                @TaskAction
+                public void renderOptionValue() {
+                    System.out.println("Value of myProp: " + myProp.getOrNull());
+                }
+
+                private static enum TestEnum {
+                    OPT_1, OPT_2, OPT_3
+                }
+            }
+        """
+    }
+
+    String groovyTaskWithListPropertyOption(String optionType) {
+        """
+            public class SampleTask extends DefaultTask {
+                @Internal
+                @Option(option = "myProp", description = "Configures command line option 'myProp'.")
+                final ListProperty<$optionType> myProp = project.objects.listProperty($optionType)
+
+                @TaskAction
+                public void renderOptionValue() {
+                    println("Value of myProp: " + myProp.getOrNull())
+                }
+
+                private static enum TestEnum {
+                    OPT_1, OPT_2, OPT_3
+                }
+            }
+        """
+    }
+
+    String taskWithSetPropertyOption(String optionType) {
+        """
+            import org.gradle.api.DefaultTask;
+            import org.gradle.api.tasks.Internal;
+            import org.gradle.api.tasks.TaskAction;
+            import org.gradle.api.tasks.options.Option;
+            import org.gradle.api.provider.SetProperty;
+
+            public class SampleTask extends DefaultTask {
+                private final SetProperty<$optionType> myProp = getProject().getObjects().setProperty(${optionType}.class);
+
+                @Internal
+                @Option(option = "myProp", description = "Configures command line option 'myProp'.")
+                public SetProperty<$optionType> getMyProp() {
+                    return myProp;
+                }
+
+                @TaskAction
+                public void renderOptionValue() {
+                    System.out.println("Value of myProp: " + myProp.getOrNull());
+                }
+
+                private static enum TestEnum {
+                    OPT_1, OPT_2, OPT_3
+                }
+            }
+        """
+    }
+
+    String groovyTaskWithSetPropertyOption(String optionType) {
+        """
+            public class SampleTask extends DefaultTask {
+                @Internal
+                @Option(option = "myProp", description = "Configures command line option 'myProp'.")
+                final SetProperty<$optionType> myProp = project.objects.setProperty($optionType)
+
+                @TaskAction
+                public void renderOptionValue() {
+                    println("Value of myProp: " + myProp.getOrNull())
+                }
+
+                private static enum TestEnum {
+                    OPT_1, OPT_2, OPT_3
+                }
+            }
+        """
+    }
+
     String taskWithMultipleOptions() {
         """
             import org.gradle.api.DefaultTask;

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/options/TaskOptionIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/options/TaskOptionIntegrationTest.groovy
@@ -133,7 +133,7 @@ Options
 
     def "can render option with help for Java task with property"() {
         given:
-        file('buildSrc/src/main/java/SampleTask.java') << taskWithSinglePropertyOption('String')
+        file('buildSrc/src/main/java/SampleTask.java') << taskWithSinglePropertyOption('Property', 'String')
         buildFile << sampleTask()
 
         when:
@@ -147,7 +147,7 @@ Options
 
     def "can render option with help for Groovy task with property"() {
         given:
-        buildFile << groovyTaskWithSinglePropertyOption('String')
+        buildFile << groovyTaskWithSinglePropertyOption('Property', 'String')
         buildFile << sampleTask()
 
         when:
@@ -200,9 +200,9 @@ Options
         outputContains("Value of myProp: fromConfigureTask")
     }
 
-    def "set value of property of type Property of type #optionType when #description for Java task"() {
+    def "set value of property of type #propertyType of type #optionType when #description for Java task"() {
         given:
-        file('buildSrc/src/main/java/SampleTask.java') << taskWithSinglePropertyOption(optionType)
+        file('buildSrc/src/main/java/SampleTask.java') << taskWithSinglePropertyOption(propertyType, optionType)
         buildFile << sampleTask()
 
         when:
@@ -212,20 +212,30 @@ Options
         outputContains("Value of myProp: $optionValue")
 
         where:
-        optionType | options                        | optionValue         | description
-        'String'   | ['--myProp=test']              | 'test'              | 'provided'
-        'String'   | ['--myProp=ab\'c=123:x\\yz45'] | 'ab\'c=123:x\\yz45' | 'provided with special characters'
-        'String'   | []                             | 'null '             | 'not provided'
-        'Boolean'  | ['--myProp']                   | 'true'              | 'provided'
-        'Boolean'  | []                             | 'null '             | 'not provided'
-        'TestEnum' | ['--myProp=OPT_2']             | 'OPT_2'             | 'provided with upper case'
-        'TestEnum' | ['--myProp=opt_2']             | 'OPT_2'             | 'provided with lower case'
-        'TestEnum' | []                             | 'null'              | 'not provided'
+        propertyType     | optionType | options                              | optionValue         | description
+        'Property'       | 'String'   | ['--myProp=test']                    | 'test'              | 'provided'
+        'Property'       | 'String'   | ['--myProp=ab\'c=123:x\\yz45']       | 'ab\'c=123:x\\yz45' | 'provided with special characters'
+        'Property'       | 'String'   | []                                   | 'null '             | 'not provided'
+        'Property'       | 'Boolean'  | ['--myProp']                         | 'true'              | 'provided'
+        'Property'       | 'Boolean'  | []                                   | 'null '             | 'not provided'
+        'Property'       | 'TestEnum' | ['--myProp=OPT_2']                   | 'OPT_2'             | 'provided with upper case'
+        'Property'       | 'TestEnum' | ['--myProp=opt_2']                   | 'OPT_2'             | 'provided with lower case'
+        'Property'       | 'TestEnum' | []                                   | 'null'              | 'not provided'
+        'ListProperty'   | 'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'            | 'provided'
+        'ListProperty'   | 'String'   | []                                   | '[]'                | 'not provided'
+        'ListProperty'   | 'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]'    | 'provided with upper case'
+        'ListProperty'   | 'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]'    | 'provided with lower case'
+        'ListProperty'   | 'TestEnum' | []                                   | '[]'                | 'not provided'
+        'SetProperty'    | 'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'            | 'provided'
+        'SetProperty'    | 'String'   | []                                   | '[]'                | 'not provided'
+        'SetProperty'    | 'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]'    | 'provided with upper case'
+        'SetProperty'    | 'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]'    | 'provided with lower case'
+        'SetProperty'    | 'TestEnum' | []                                   | '[]'                | 'not provided'
     }
 
-    def "set value of property of type Property of type #optionType when #description for Groovy task"() {
+    def "set value of property of type #propertyType of type #optionType when #description for Groovy task"() {
         given:
-        buildFile << groovyTaskWithSinglePropertyOption(optionType)
+        buildFile << groovyTaskWithSinglePropertyOption(propertyType, optionType)
         buildFile << sampleTask()
 
         when:
@@ -235,95 +245,25 @@ Options
         outputContains("Value of myProp: $optionValue")
 
         where:
-        optionType | options                        | optionValue         | description
-        'String'   | ['--myProp=test']              | 'test'              | 'provided'
-        'String'   | ['--myProp=ab\'c=123:x\\yz45'] | 'ab\'c=123:x\\yz45' | 'provided with special characters'
-        'String'   | []                             | 'null '             | 'not provided'
-        'Boolean'  | ['--myProp']                   | 'true'              | 'provided'
-        'Boolean'  | []                             | 'null '             | 'not provided'
-        'TestEnum' | ['--myProp=OPT_2']             | 'OPT_2'             | 'provided with upper case'
-        'TestEnum' | ['--myProp=opt_2']             | 'OPT_2'             | 'provided with lower case'
-        'TestEnum' | []                             | 'null'              | 'not provided'
-    }
-
-    def "set value of property of type ListProperty of type #optionType when #description for Java task"() {
-        given:
-        file('buildSrc/src/main/java/SampleTask.java') << taskWithListPropertyOption(optionType)
-        buildFile << sampleTask()
-
-        when:
-        run(['sample'] + options as String[])
-
-        then:
-        outputContains("Value of myProp: $optionValue")
-
-        where:
-        optionType | options                              | optionValue        | description
-        'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'           | 'provided'
-        'String'   | []                                   | '[]'               | 'not provided'
-        'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]'   | 'provided with upper case'
-        'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]'   | 'provided with lower case'
-        'TestEnum' | []                                   | '[]'               | 'not provided'
-    }
-
-    def "set value of property of type ListProperty of type #optionType when #description for Groovy task"() {
-        given:
-        buildFile << groovyTaskWithListPropertyOption(optionType)
-        buildFile << sampleTask()
-
-        when:
-        run(['sample'] + options as String[])
-
-        then:
-        outputContains("Value of myProp: $optionValue")
-
-        where:
-        optionType | options                              | optionValue      | description
-        'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'         | 'provided'
-        'String'   | []                                   | '[]'             | 'not provided'
-        'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]' | 'provided with upper case'
-        'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]' | 'provided with lower case'
-        'TestEnum' | []                                   | '[]'             | 'not provided'
-    }
-
-    def "set value of property of type SetProperty of type #optionType when #description for Java task"() {
-        given:
-        file('buildSrc/src/main/java/SampleTask.java') << taskWithSetPropertyOption(optionType)
-        buildFile << sampleTask()
-
-        when:
-        run(['sample'] + options as String[])
-
-        then:
-        outputContains("Value of myProp: $optionValue")
-
-        where:
-        optionType | options                              | optionValue        | description
-        'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'           | 'provided'
-        'String'   | []                                   | '[]'               | 'not provided'
-        'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]'   | 'provided with upper case'
-        'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]'   | 'provided with lower case'
-        'TestEnum' | []                                   | '[]'               | 'not provided'
-    }
-
-    def "set value of property of type SetProperty of type #optionType when #description for Groovy task"() {
-        given:
-        buildFile << groovyTaskWithSetPropertyOption(optionType)
-        buildFile << sampleTask()
-
-        when:
-        run(['sample'] + options as String[])
-
-        then:
-        outputContains("Value of myProp: $optionValue")
-
-        where:
-        optionType | options                              | optionValue      | description
-        'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'         | 'provided'
-        'String'   | []                                   | '[]'             | 'not provided'
-        'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]' | 'provided with upper case'
-        'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]' | 'provided with lower case'
-        'TestEnum' | []                                   | '[]'             | 'not provided'
+        propertyType     | optionType | options                              | optionValue         | description
+        'Property'       |'String'    | ['--myProp=test']                    | 'test'              | 'provided'
+        'Property'       |'String'    | ['--myProp=ab\'c=123:x\\yz45']       | 'ab\'c=123:x\\yz45' | 'provided with special characters'
+        'Property'       |'String'    | []                                   | 'null '             | 'not provided'
+        'Property'       |'Boolean'   | ['--myProp']                         | 'true'              | 'provided'
+        'Property'       |'Boolean'   | []                                   | 'null '             | 'not provided'
+        'Property'       |'TestEnum'  | ['--myProp=OPT_2']                   | 'OPT_2'             | 'provided with upper case'
+        'Property'       |'TestEnum'  | ['--myProp=opt_2']                   | 'OPT_2'             | 'provided with lower case'
+        'Property'       |'TestEnum'  | []                                   | 'null'              | 'not provided'
+        'ListProperty'   | 'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'            | 'provided'
+        'ListProperty'   | 'String'   | []                                   | '[]'                | 'not provided'
+        'ListProperty'   | 'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]'    | 'provided with upper case'
+        'ListProperty'   | 'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]'    | 'provided with lower case'
+        'ListProperty'   | 'TestEnum' | []                                   | '[]'                | 'not provided'
+        'SetProperty'    | 'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'            | 'provided'
+        'SetProperty'    | 'String'   | []                                   | '[]'                | 'not provided'
+        'SetProperty'    | 'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]'    | 'provided with upper case'
+        'SetProperty'    | 'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]'    | 'provided with lower case'
+        'SetProperty'    | 'TestEnum' | []                                   | '[]'                | 'not provided'
     }
 
     static String sampleTask() {

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/options/TaskOptionIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/options/TaskOptionIntegrationTest.groovy
@@ -246,6 +246,86 @@ Options
         'TestEnum' | []                             | 'null'              | 'not provided'
     }
 
+    def "set value of property of type ListProperty of type #optionType when #description for Java task"() {
+        given:
+        file('buildSrc/src/main/java/SampleTask.java') << taskWithListPropertyOption(optionType)
+        buildFile << sampleTask()
+
+        when:
+        run(['sample'] + options as String[])
+
+        then:
+        outputContains("Value of myProp: $optionValue")
+
+        where:
+        optionType | options                              | optionValue        | description
+        'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'           | 'provided'
+        'String'   | []                                   | '[]'               | 'not provided'
+        'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]'   | 'provided with upper case'
+        'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]'   | 'provided with lower case'
+        'TestEnum' | []                                   | '[]'               | 'not provided'
+    }
+
+    def "set value of property of type ListProperty of type #optionType when #description for Groovy task"() {
+        given:
+        buildFile << groovyTaskWithListPropertyOption(optionType)
+        buildFile << sampleTask()
+
+        when:
+        run(['sample'] + options as String[])
+
+        then:
+        outputContains("Value of myProp: $optionValue")
+
+        where:
+        optionType | options                              | optionValue      | description
+        'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'         | 'provided'
+        'String'   | []                                   | '[]'             | 'not provided'
+        'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]' | 'provided with upper case'
+        'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]' | 'provided with lower case'
+        'TestEnum' | []                                   | '[]'             | 'not provided'
+    }
+
+    def "set value of property of type SetProperty of type #optionType when #description for Java task"() {
+        given:
+        file('buildSrc/src/main/java/SampleTask.java') << taskWithSetPropertyOption(optionType)
+        buildFile << sampleTask()
+
+        when:
+        run(['sample'] + options as String[])
+
+        then:
+        outputContains("Value of myProp: $optionValue")
+
+        where:
+        optionType | options                              | optionValue        | description
+        'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'           | 'provided'
+        'String'   | []                                   | '[]'               | 'not provided'
+        'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]'   | 'provided with upper case'
+        'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]'   | 'provided with lower case'
+        'TestEnum' | []                                   | '[]'               | 'not provided'
+    }
+
+    def "set value of property of type SetProperty of type #optionType when #description for Groovy task"() {
+        given:
+        buildFile << groovyTaskWithSetPropertyOption(optionType)
+        buildFile << sampleTask()
+
+        when:
+        run(['sample'] + options as String[])
+
+        then:
+        outputContains("Value of myProp: $optionValue")
+
+        where:
+        optionType | options                              | optionValue      | description
+        'String'   | ['--myProp=a', '--myProp=b']         | '[a, b]'         | 'provided'
+        'String'   | []                                   | '[]'             | 'not provided'
+        'TestEnum' | ['--myProp=OPT_1', '--myProp=OPT_2'] | '[OPT_1, OPT_2]' | 'provided with upper case'
+        'TestEnum' | ['--myProp=opt_1', '--myProp=opt_2'] | '[OPT_1, OPT_2]' | 'provided with lower case'
+        'TestEnum' | []                                   | '[]'             | 'not provided'
+    }
+
     static String sampleTask() {
         """
             task sample(type: SampleTask)

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/FieldOptionElement.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/FieldOptionElement.java
@@ -18,9 +18,8 @@ package org.gradle.api.internal.tasks.options;
 
 import org.apache.commons.lang.StringUtils;
 import org.gradle.api.NonNullApi;
-import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.HasMultipleValues;
 import org.gradle.api.provider.Property;
-import org.gradle.api.provider.SetProperty;
 import org.gradle.api.tasks.options.Option;
 import org.gradle.internal.Cast;
 import org.gradle.internal.reflect.JavaMethod;
@@ -40,7 +39,7 @@ public class FieldOptionElement {
             PropertySetter setter = mutateUsingGetter(field);
             return AbstractOptionElement.of(optionName, option, setter, optionValueNotationParserFactory);
         }
-        if (ListProperty.class.isAssignableFrom(fieldType) || SetProperty.class.isAssignableFrom(fieldType)) {
+        if (HasMultipleValues.class.isAssignableFrom(fieldType)) {
             PropertySetter setter = mutateUsingGetter(field);
             Class<?> elementType = setter.getRawType();
             return new MultipleValueOptionElement(optionName, option, elementType, setter, optionValueNotationParserFactory);
@@ -55,11 +54,8 @@ public class FieldOptionElement {
     }
 
     private static PropertySetter mutateUsingGetter(final Field field) {
-        if (ListProperty.class.isAssignableFrom(field.getType())) {
-            return new ListPropertyFieldSetter(getGetter(field), field);
-        }
-        if (SetProperty.class.isAssignableFrom(field.getType())) {
-            return new SetPropertyFieldSetter(getGetter(field), field);
+        if (HasMultipleValues.class.isAssignableFrom(field.getType())) {
+            return new MultipleValuePropertyFieldSetter(getGetter(field), field);
         }
         return new PropertyFieldSetter(getGetter(field), field);
     }
@@ -160,27 +156,14 @@ public class FieldOptionElement {
     }
 
     @NonNullApi
-    private static class ListPropertyFieldSetter extends PropertyFieldSetter {
-        public ListPropertyFieldSetter(Method getter, Field field) {
+    private static class MultipleValuePropertyFieldSetter extends PropertyFieldSetter {
+        public MultipleValuePropertyFieldSetter(Method getter, Field field) {
             super(getter, field);
         }
 
         @Override
         public void setValue(Object target, Object value) {
-            ListProperty<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, getGetter()).invoke(target));
-            property.set((Iterable<?>) value);
-        }
-    }
-
-    @NonNullApi
-    private static class SetPropertyFieldSetter extends PropertyFieldSetter {
-        public SetPropertyFieldSetter(Method getter, Field field) {
-            super(getter, field);
-        }
-
-        @Override
-        public void setValue(Object target, Object value) {
-            SetProperty<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, getGetter()).invoke(target));
+            HasMultipleValues<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, getGetter()).invoke(target));
             property.set((Iterable<?>) value);
         }
     }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/FieldOptionElement.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/FieldOptionElement.java
@@ -17,7 +17,10 @@
 package org.gradle.api.internal.tasks.options;
 
 import org.apache.commons.lang.StringUtils;
+import org.gradle.api.NonNullApi;
+import org.gradle.api.provider.ListProperty;
 import org.gradle.api.provider.Property;
+import org.gradle.api.provider.SetProperty;
 import org.gradle.api.tasks.options.Option;
 import org.gradle.internal.Cast;
 import org.gradle.internal.reflect.JavaMethod;
@@ -37,6 +40,11 @@ public class FieldOptionElement {
             PropertySetter setter = mutateUsingGetter(field);
             return AbstractOptionElement.of(optionName, option, setter, optionValueNotationParserFactory);
         }
+        if (ListProperty.class.isAssignableFrom(fieldType) || SetProperty.class.isAssignableFrom(fieldType)) {
+            PropertySetter setter = mutateUsingGetter(field);
+            Class<?> elementType = setter.getRawType();
+            return new MultipleValueOptionElement(optionName, option, elementType, setter, optionValueNotationParserFactory);
+        }
 
         PropertySetter setter = mutateUsingSetter(field);
         return AbstractOptionElement.of(optionName, option, setter, optionValueNotationParserFactory);
@@ -47,6 +55,12 @@ public class FieldOptionElement {
     }
 
     private static PropertySetter mutateUsingGetter(final Field field) {
+        if (ListProperty.class.isAssignableFrom(field.getType())) {
+            return new ListPropertyFieldSetter(getGetter(field), field);
+        }
+        if (SetProperty.class.isAssignableFrom(field.getType())) {
+            return new SetPropertyFieldSetter(getGetter(field), field);
+        }
         return new PropertyFieldSetter(getGetter(field), field);
     }
 
@@ -138,6 +152,36 @@ public class FieldOptionElement {
         public void setValue(Object target, Object value) {
             Property<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, getter).invoke(target));
             property.set(value);
+        }
+
+        protected Method getGetter() {
+            return getter;
+        }
+    }
+
+    @NonNullApi
+    private static class ListPropertyFieldSetter extends PropertyFieldSetter {
+        public ListPropertyFieldSetter(Method getter, Field field) {
+            super(getter, field);
+        }
+
+        @Override
+        public void setValue(Object target, Object value) {
+            ListProperty<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, getGetter()).invoke(target));
+            property.set((Iterable<?>) value);
+        }
+    }
+
+    @NonNullApi
+    private static class SetPropertyFieldSetter extends PropertyFieldSetter {
+        public SetPropertyFieldSetter(Method getter, Field field) {
+            super(getter, field);
+        }
+
+        @Override
+        public void setValue(Object target, Object value) {
+            SetProperty<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, getGetter()).invoke(target));
+            property.set((Iterable<?>) value);
         }
     }
 }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/MethodOptionElement.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/MethodOptionElement.java
@@ -17,9 +17,8 @@
 package org.gradle.api.internal.tasks.options;
 
 import org.gradle.api.NonNullApi;
-import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.HasMultipleValues;
 import org.gradle.api.provider.Property;
-import org.gradle.api.provider.SetProperty;
 import org.gradle.api.tasks.options.Option;
 import org.gradle.internal.Cast;
 import org.gradle.internal.reflect.JavaMethod;
@@ -44,7 +43,7 @@ public class MethodOptionElement {
             PropertySetter setter = mutateUsingReturnValue(method);
             return AbstractOptionElement.of(optionName, option, setter, optionValueNotationParserFactory);
         }
-        if (ListProperty.class.isAssignableFrom(method.getReturnType()) || SetProperty.class.isAssignableFrom(method.getReturnType())) {
+        if (HasMultipleValues.class.isAssignableFrom(method.getReturnType())) {
             assertCanUseMethodReturnType(optionName, method);
             PropertySetter setter = mutateUsingReturnValue(method);
             Class<?> elementType = setter.getRawType();
@@ -68,11 +67,8 @@ public class MethodOptionElement {
     }
 
     private static PropertySetter mutateUsingReturnValue(Method method) {
-        if (ListProperty.class.isAssignableFrom(method.getReturnType())) {
-            return new ListPropertyValueSetter(method);
-        }
-        if (SetProperty.class.isAssignableFrom(method.getReturnType())) {
-            return new SetPropertyValueSetter(method);
+        if (HasMultipleValues.class.isAssignableFrom(method.getReturnType())) {
+            return new MultipleValuePropertyValueSetter(method);
         }
         return new PropertyValueSetter(method);
     }
@@ -157,27 +153,14 @@ public class MethodOptionElement {
     }
 
     @NonNullApi
-    private static class ListPropertyValueSetter extends PropertyValueSetter {
-        public ListPropertyValueSetter(Method method) {
+    private static class MultipleValuePropertyValueSetter extends PropertyValueSetter {
+        public MultipleValuePropertyValueSetter(Method method) {
             super(method);
         }
 
         @Override
         public void setValue(Object target, Object value) {
-            ListProperty<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, getMethod()).invoke(target));
-            property.set((Iterable<?>) value);
-        }
-    }
-
-    @NonNullApi
-    private static class SetPropertyValueSetter extends PropertyValueSetter {
-        public SetPropertyValueSetter(Method method) {
-            super(method);
-        }
-
-        @Override
-        public void setValue(Object target, Object value) {
-            SetProperty<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, getMethod()).invoke(target));
+            HasMultipleValues<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, getMethod()).invoke(target));
             property.set((Iterable<?>) value);
         }
     }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/MethodOptionElement.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/MethodOptionElement.java
@@ -16,7 +16,10 @@
 
 package org.gradle.api.internal.tasks.options;
 
+import org.gradle.api.NonNullApi;
+import org.gradle.api.provider.ListProperty;
 import org.gradle.api.provider.Property;
+import org.gradle.api.provider.SetProperty;
 import org.gradle.api.tasks.options.Option;
 import org.gradle.internal.Cast;
 import org.gradle.internal.reflect.JavaMethod;
@@ -41,6 +44,12 @@ public class MethodOptionElement {
             PropertySetter setter = mutateUsingReturnValue(method);
             return AbstractOptionElement.of(optionName, option, setter, optionValueNotationParserFactory);
         }
+        if (ListProperty.class.isAssignableFrom(method.getReturnType()) || SetProperty.class.isAssignableFrom(method.getReturnType())) {
+            assertCanUseMethodReturnType(optionName, method);
+            PropertySetter setter = mutateUsingReturnValue(method);
+            Class<?> elementType = setter.getRawType();
+            return new MultipleValueOptionElement(optionName, option, elementType, setter, optionValueNotationParserFactory);
+        }
         if (method.getParameterTypes().length == 0) {
             return new BooleanOptionElement(optionName, option, setFlagUsingMethod(method));
         }
@@ -59,6 +68,12 @@ public class MethodOptionElement {
     }
 
     private static PropertySetter mutateUsingReturnValue(Method method) {
+        if (ListProperty.class.isAssignableFrom(method.getReturnType())) {
+            return new ListPropertyValueSetter(method);
+        }
+        if (SetProperty.class.isAssignableFrom(method.getReturnType())) {
+            return new SetPropertyValueSetter(method);
+        }
         return new PropertyValueSetter(method);
     }
 
@@ -134,6 +149,36 @@ public class MethodOptionElement {
         public void setValue(Object target, Object value) {
             Property<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, method).invoke(target));
             property.set(value);
+        }
+
+        protected Method getMethod() {
+            return method;
+        }
+    }
+
+    @NonNullApi
+    private static class ListPropertyValueSetter extends PropertyValueSetter {
+        public ListPropertyValueSetter(Method method) {
+            super(method);
+        }
+
+        @Override
+        public void setValue(Object target, Object value) {
+            ListProperty<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, getMethod()).invoke(target));
+            property.set((Iterable<?>) value);
+        }
+    }
+
+    @NonNullApi
+    private static class SetPropertyValueSetter extends PropertyValueSetter {
+        public SetPropertyValueSetter(Method method) {
+            super(method);
+        }
+
+        @Override
+        public void setValue(Object target, Object value) {
+            SetProperty<Object> property = Cast.uncheckedNonnullCast(JavaMethod.of(Object.class, getMethod()).invoke(target));
+            property.set((Iterable<?>) value);
         }
     }
 

--- a/subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy
@@ -17,7 +17,9 @@
 package org.gradle.api.internal.tasks.options
 
 import org.gradle.api.internal.tasks.TaskOptionsGenerator
+import org.gradle.api.provider.ListProperty
 import org.gradle.api.provider.Property
+import org.gradle.api.provider.SetProperty
 import org.gradle.api.tasks.options.Option
 import org.gradle.api.tasks.options.OptionValues
 import spock.lang.Issue
@@ -106,6 +108,32 @@ class OptionReaderTest extends Specification {
         options[4].description == "string value"
         options[4].argumentType == String
         options[4].availableValues == ["dynValue1", "dynValue2"] as Set
+    }
+
+    def "can read options linked to property getter methods of type ListProperty of a task"() {
+        when:
+        List<InstanceOptionDescriptor> options = TaskOptionsGenerator.generate(new TestClassWithListProperties(), reader).getAll()
+        then:
+        options[0].name == "enumValue"
+        options[0].description == "enum value"
+        options[0].argumentType == List
+
+        options[1].name == "stringValue"
+        options[1].description == "string value"
+        options[1].argumentType == List
+    }
+
+    def "can read options linked to property getter methods of type SetProperty of a task"() {
+        when:
+        List<InstanceOptionDescriptor> options = TaskOptionsGenerator.generate(new TestClassWithSetProperties(), reader).getAll()
+        then:
+        options[0].name == "enumValue"
+        options[0].description == "enum value"
+        options[0].argumentType == List
+
+        options[1].name == "stringValue"
+        options[1].description == "string value"
+        options[1].argumentType == List
     }
 
     def "built-in options appear last"() {
@@ -274,6 +302,31 @@ class OptionReaderTest extends Specification {
         options[3].availableValues.isEmpty()
     }
 
+    def "handles property field options of type ListProperty"() {
+        when:
+        List<InstanceOptionDescriptor> options = TaskOptionsGenerator.generate(new TestClassWithListPropertyField(), reader).getAll()
+        then:
+        options[0].name == "customOptionName"
+        options[0].description == "custom description"
+        options[0].argumentType == List
+
+        options[1].name == "field2"
+        options[1].description == "Descr Field2"
+        options[1].argumentType == List
+    }
+
+    def "handles property field options of type SetProperty"() {
+        when:
+        List<InstanceOptionDescriptor> options = TaskOptionsGenerator.generate(new TestClassWithSetPropertyField(), reader).getAll()
+        then:
+        options[0].name == "customOptionName"
+        options[0].description == "custom description"
+        options[0].argumentType == List
+
+        options[1].name == "field2"
+        options[1].description == "Descr Field2"
+        options[1].argumentType == List
+    }
 
     def "throws decent error when description not set"() {
         when:
@@ -527,6 +580,30 @@ class OptionReaderTest extends Specification {
         }
     }
 
+    public static class TestClassWithListProperties {
+        @Option(option = "stringValue", description = "string value")
+        public ListProperty<String> getStringValue() {
+            throw new UnsupportedOperationException()
+        }
+
+        @Option(option = "enumValue", description = "enum value")
+        public ListProperty<TestEnum> getEnumValue() {
+            throw new UnsupportedOperationException()
+        }
+    }
+
+    public static class TestClassWithSetProperties {
+        @Option(option = "stringValue", description = "string value")
+        public SetProperty<String> getStringValue() {
+            throw new UnsupportedOperationException()
+        }
+
+        @Option(option = "enumValue", description = "enum value")
+        public SetProperty<TestEnum> getEnumValue() {
+            throw new UnsupportedOperationException()
+        }
+    }
+
     public static class CustomClass {
         String value
 
@@ -629,6 +706,22 @@ class OptionReaderTest extends Specification {
         List<String> getField2Options() {
             return Arrays.asList("dynValue1", "dynValue2")
         }
+    }
+
+    public static class TestClassWithListPropertyField {
+        @Option(option = 'customOptionName', description = "custom description")
+        final ListProperty<String> field1
+
+        @Option(description = "Descr Field2")
+        final ListProperty<TestEnum> field2
+    }
+
+    public static class TestClassWithSetPropertyField {
+        @Option(option = 'customOptionName', description = "custom description")
+        final SetProperty<String> field1
+
+        @Option(description = "Descr Field2")
+        final SetProperty<TestEnum> field2
     }
 
     public static class TestClass7 {

--- a/subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy
@@ -302,7 +302,7 @@ class OptionReaderTest extends Specification {
         options[3].availableValues.isEmpty()
 
         options[4].name == "no-field4"
-        options[4].description == "Disables option --field4"
+        options[4].description == "Disables option --field4."
         options[4].argumentType == Void.TYPE
         options[4].availableValues.isEmpty()
 

--- a/subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy
@@ -300,32 +300,27 @@ class OptionReaderTest extends Specification {
         options[3].description == "Descr Field4"
         options[3].argumentType == Void.TYPE
         options[3].availableValues.isEmpty()
-    }
 
-    def "handles property field options of type ListProperty"() {
-        when:
-        List<InstanceOptionDescriptor> options = TaskOptionsGenerator.generate(new TestClassWithListPropertyField(), reader).getAll()
-        then:
-        options[0].name == "customOptionName"
-        options[0].description == "custom description"
-        options[0].argumentType == List
+        options[4].name == "no-field4"
+        options[4].description == "Disables option --field4"
+        options[4].argumentType == Void.TYPE
+        options[4].availableValues.isEmpty()
 
-        options[1].name == "field2"
-        options[1].description == "Descr Field2"
-        options[1].argumentType == List
-    }
+        options[5].name == "field5"
+        options[5].description == "Descr Field5"
+        options[5].argumentType == List
 
-    def "handles property field options of type SetProperty"() {
-        when:
-        List<InstanceOptionDescriptor> options = TaskOptionsGenerator.generate(new TestClassWithSetPropertyField(), reader).getAll()
-        then:
-        options[0].name == "customOptionName"
-        options[0].description == "custom description"
-        options[0].argumentType == List
+        options[6].name == "field6"
+        options[6].description == "Descr Field6"
+        options[6].argumentType == List
 
-        options[1].name == "field2"
-        options[1].description == "Descr Field2"
-        options[1].argumentType == List
+        options[7].name == "field7"
+        options[7].description == "Descr Field7"
+        options[7].argumentType == List
+
+        options[8].name == "field8"
+        options[8].description == "Descr Field8"
+        options[8].argumentType == List
     }
 
     def "throws decent error when description not set"() {
@@ -702,26 +697,22 @@ class OptionReaderTest extends Specification {
         @Option(description = "Descr Field4")
         final Property<Boolean> field4
 
+        @Option(description = "Descr Field5")
+        final ListProperty<String> field5
+
+        @Option(description = "Descr Field6")
+        final ListProperty<TestEnum> field6
+
+        @Option(description = "Descr Field7")
+        final SetProperty<String> field7
+
+        @Option(description = "Descr Field8")
+        final SetProperty<TestEnum> field8
+
         @OptionValues("field2")
         List<String> getField2Options() {
             return Arrays.asList("dynValue1", "dynValue2")
         }
-    }
-
-    public static class TestClassWithListPropertyField {
-        @Option(option = 'customOptionName', description = "custom description")
-        final ListProperty<String> field1
-
-        @Option(description = "Descr Field2")
-        final ListProperty<TestEnum> field2
-    }
-
-    public static class TestClassWithSetPropertyField {
-        @Option(option = 'customOptionName', description = "custom description")
-        final SetProperty<String> field1
-
-        @Option(description = "Descr Field2")
-        final SetProperty<TestEnum> field2
     }
 
     public static class TestClass7 {

